chainsaw-demo/src/grpc_impl.rs

```diff
@@ -11,7 +11,7 @@ impl Greeter for MyGreeter {
         &self,
         request: Request<HelloRequest>,
     ) -> Result<Response<HelloReply>, Status> {
-        info!("Handling request.");
+        tracing::info!("Handling request.");
```

Comment:
Is this a style change? Or is a side effect of moving into a `bin` directory affecting the scope of:
```rust
#[macro_use]
extern crate tracing;

```

Reply:
It's a side-effect of the bin being a separate crate now, yeah. Since `extern crate ...` is no longer necessary to consume macros, I'm not sure whether it's more idiomatic to explicitly reference the crate they're from or not, particularly as using lots of macros makes it difficult to figure out which crate they come from.

Reply:
I suppose there are three ways of going about it:
1. `#[macro_use] extern crate tracing;`
2. `use tracing::{info, warn, debug, error};` in multiple modules
3. Fully qualified `tracing::info!(...); tracing::error!(...)`
I like the first one because it means logging is easy for the developer, but as you said it's not clear where the macro comes from. The foot-gun there is that the scope effects of `#[macro_use]` aren't exactly clear, and there's potential you end up reverting back to the standard `log` facade without realising, and you wonder why your structured logging isn't working.

Number 2 works in the same way and makes the scope of the import obvious, though it can clutter up the top of your module, and you might use something like `debug!(...)` without realising if it's actually imported further up.

Finally, fully qualifying makes it obvious to the developer but introduces some friction due to the `tracing::` prefix whenever you want to log.

After listing those out fully qualifying feels like it's more idiomatic to me, as there's no risk of hidden scope issues or forgetting to add an import, but I'm not sure how you feel on it.

Reply:
To compare it against Go, we have:

```go
logger.Error("whatever")
```

For Python we have:

```python
logger.error("whatever")
```

Having a fully-qualified prefix of `tracing::error!("whatever")` doesn't seem too jarring compared to those, and I like the lack of hidden scope in the general spirit of being obvious and avoiding gotchas.